common/src/chain/transaction/signature/tests/sign_and_mutate.rs

```diff
@@ -0,0 +1,509 @@
+        add_output,
+        modify_output,
+        remove_first_output,
+        remove_middle_output,
+        remove_last_output,
+    ];
+    check_mutations(&tx, &destination, mutations, Ok(()));
+}
+
+#[test]
+fn mutate_single() {
+    let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+    let destination = Destination::PublicKey(public_key);
+    let sighash_type = SigHashType::try_from(SigHashType::SINGLE).unwrap();
+    let tx =
+        generate_and_sigh_tx(&destination, INPUTS, OUTPUTS, &private_key, sighash_type).unwrap();
+
+    let mutations = [
+        add_input,
+        modify_input,
+        remove_first_input,
+        remove_middle_input,
+        remove_last_input,
+        remove_first_output,
+    ];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
+            assert_eq!(
+                verify_signature(&destination, &tx, input),
+                Err(TransactionSigError::SignatureVerificationFailed)
+            );
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+
+    let mutations = [add_output, remove_last_output];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
+            assert_eq!(verify_signature(&destination, &tx, input), Ok(()));
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+
+    {
+        let tx = modify_output(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        assert_eq!(
+            verify_signature(&destination, &tx, 0),
+            Err(TransactionSigError::SignatureVerificationFailed),
+        );
+        for input in 1..inputs - 1 {
+            assert_eq!(verify_signature(&destination, &tx, input), Ok(()),);
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+}
+
+#[test]
+fn mutate_single_anyonecanpay() {
+    let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+    let destination = Destination::PublicKey(public_key);
+    let sighash_type =
+        SigHashType::try_from(SigHashType::SINGLE | SigHashType::ANYONECANPAY).unwrap();
+    let tx =
+        generate_and_sigh_tx(&destination, INPUTS, OUTPUTS, &private_key, sighash_type).unwrap();
+
+    let mutations = [add_input, remove_last_input, add_output, remove_last_output];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
```

Comment:
why inputs - 1? Is this an off-by-one error? If not, please explain it. Either in a comment, or use an iterator with `take` so that the reader can see that it's deliberate.